usersession/client/client.go

```diff
@@ -0,0 +1,268 @@
+			}
+			httpResp, err := client.doer.Do(req)
+			if err != nil {
+				response.err = err
+				return
+			}
+			defer httpResp.Body.Close()
+			response.statusCode = httpResp.StatusCode
+			response.err = decodeInto(httpResp.Body, &response)
+			response.checkError()
+		}(socket)
+	}
+	wg.Wait()
+	return responses, nil
+}
+
+func decodeInto(reader io.Reader, v interface{}) error {
+	dec := json.NewDecoder(reader)
+	if err := dec.Decode(v); err != nil {
+		r := dec.Buffered()
+		buf, err1 := ioutil.ReadAll(r)
+		if err1 != nil {
+			buf = []byte(fmt.Sprintf("error reading buffered response body: %s", err1))
+		}
+		return fmt.Errorf("cannot decode %q: %s", buf, err)
+	}
+	return nil
+}
+
+type SessionInfo struct {
+	Version string `json:"version"`
+}
+
+func (client *Client) SessionInfo(ctx context.Context) (info map[int]SessionInfo, err error) {
+	responses, err := client.doMany(ctx, "GET", "/v1/session-info", nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	info = make(map[int]SessionInfo)
+	for _, resp := range responses {
+		if resp.err != nil {
+			if err == nil {
+				err = resp.err
+			}
+			continue
+		}
+		var si SessionInfo
+		if decodeErr := json.Unmarshal(resp.Result, &si); decodeErr != nil {
+			if err == nil {
+				err = decodeErr
+			}
+			continue
+		}
+		info[resp.uid] = si
+	}
+	return info, err
+}
+
+type ServiceFailure struct {
+	Uid     int
+	Service string
+	Error   string
+}
+
+func decodeServiceErrors(uid int, errorValue map[string]interface{}, kind string) []ServiceFailure {
+	errors, ok := errorValue[kind].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+	var failures []ServiceFailure
+	for service, reason := range errors {
+		if reasonString, ok := reason.(string); ok {
+			failures = append(failures, ServiceFailure{
+				Uid:     uid,
+				Service: service,
+				Error:   reasonString,
+			})
+		} else {
+			logger.Noticef("Could not decode %s failure for %q: expected string, but got %T", kind, service, reason)
+		}
+	}
+	return failures
+}
+
+func (client *Client) serviceControlCall(ctx context.Context, action string, services []string) (startFailures, stopFailures []ServiceFailure, err error) {
+	headers := map[string]string{"Content-Type": "application/json"}
+	reqBody, err := json.Marshal(map[string]interface{}{
+		"action":   action,
+		"services": services,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	responses, err := client.doMany(ctx, "POST", "/v1/service-control", nil, headers, reqBody)
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, resp := range responses {
+		if resp.err == nil {
+			continue
+		}
+		if agentErr, ok := resp.err.(*Error); ok && agentErr.Kind == "service-control" {
+			if errorValue, ok := agentErr.Value.(map[string]interface{}); ok {
+				startFailures = append(startFailures, decodeServiceErrors(resp.uid, errorValue, "start-errors")...)
+				stopFailures = append(stopFailures, decodeServiceErrors(resp.uid, errorValue, "stop-errors")...)
```

Comment:
then if the decodeServiceErrors return an error this code should just end up using the path below with using resp.err as is